lib/src/internal/http/http_handler.dart

```diff
@@ -78,25 +78,14 @@ class HttpHandler {
     }
 
     // Execute request
-    try {
-      currentBucket?.addInFlightRequest(request);
-      final response = await httpClient.send(await request.prepareRequest());
-      currentBucket?.removeInFlightRequest(request);
-      currentBucket = _upsertBucket(request, response);
-      return _handle(request, response);
-    } on HttpClientException catch (e) {
-      currentBucket?.removeInFlightRequest(request);
-      if (e.response == null) {
-        logger.warning("Http Error on endpoint: ${request.uri}. Error: [${e.message.toString()}].");
-        rethrow;
-      }
-
-      final response = e.response as http.StreamedResponse;
-      _upsertBucket(request, response);
-
-      // Return http error
-      return _handle(request, response);
-    }
+    currentBucket?.addInFlightRequest(request);
+    final response = await client.options.httpRetryOptions.retry(
+      () async => httpClient.send(await request.prepareRequest()),
+      onRetry: (ex) => logger.shout('Exception when sending HTTP request (retrying automatically): $ex'),
+    );
```

Comment:
shoudn't it be implemented similarly to package example, that we retry only on TimeoutException or similar?

Reply:
We should retry on any exception, which include `TImeoutException` but also other common exceptions like `SocketException`, most of which are wrapped in `ClientException` by the http package anyways.

Reply:
We should retry on any exception, which include `TImeoutException` but also other common exceptions like `SocketException`, most of which are wrapped in `ClientException` by the http package anyways.

Reply:
what if our out data we send is wrong (server responds with 400)? It will fail anyway but this time ~10 times

Reply:
A 4xx response code is still considered a success by the http package. It's up to us to determine if the response is an error or not, which is handled in `HttpHandler._handle` (converting the response from the http package into an `HttpResponseSuccess` or `HttpResponseError` depending on it's status code) and in `HttpEndpoints.executeSafe`, which throes if the response is not an `HttpResponseSuccess`).

This function will only retry if the http package throws an exception, not if we throw one as that is handled elsewhere. So no, as long as the request is correctly sent and received (even if it represents an error) then this function will not retry.

See the docs added for `ClientOptions` which also mention this.

Reply:
i think now we consider HttpResponseSuccess if status code is >= 200 < 300. I think we should extend this behavior to correctly handle connection problems/timeouts since retrying on 400 doesnt make sense (and now it will try to retry on 400 since we are throwing exception if response is not HttpResponseSuccess

Reply:
No - `HttpResponse` from nyxx isn't in the picture here. Only the `http` package, which only throws on actual errors and not responses with a 4xx response code.

Throwing `HttpResponseError`s is done elsewhere using the return value of this function. Since the place where that is handled is *not* wrapped in retry, that will not trigger a retry.

As long as the server responds, even with an erroneous code, an exception will not be thrown here and this will not retry,

Reply:
yeah sure i do understand that. But we cant implement that feature without excluding HTTP 400 s for example because its gonna try send bad data for example 10 times. We have to do something with that. Either by wrapping http behavior and throwing expection or doing something else to go around that problem